pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/advisor.go

```diff
@@ -121,13 +117,52 @@ func NewCPUResourceAdvisor(conf *config.Configuration, extraConf interface{}, me
 	return cra
 }
 
-func (cra *cpuResourceAdvisor) Name() string {
-	return cra.name
+func (cra *cpuResourceAdvisor) Run(ctx context.Context) {
+	for {
+		select {
+		case <-cra.recvCh:
+			klog.Infof("[qosaware-cpu] receive update trigger from cpu server")
+			cra.update()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (cra *cpuResourceAdvisor) GetChannels() (interface{}, interface{}) {
+	return cra.recvCh, cra.sendCh
 }
 
-// Update works in a monolith way to maintain lifecycle and trigger update actions for all regions;
-// todo, re-consider whether it's efficient or we should make start individual goroutine for each region
-func (cra *cpuResourceAdvisor) Update() {
+func (cra *cpuResourceAdvisor) GetHeadroom() (resource.Quantity, error) {
+	cra.mutex.RLock()
+	defer cra.mutex.RUnlock()
+
+	reservePoolSize, ok := cra.metaCache.GetPoolSize(state.PoolNameReserve)
+	if !ok {
+		return resource.Quantity{}, fmt.Errorf("reserve pool not exist")
+	}
+
+	// Return maximum available resource value as headroom when no region exists
+	if len(cra.regionMap) <= 0 {
+		return *resource.NewQuantity(int64(cra.metaServer.NumCPUs-reservePoolSize), resource.DecimalSI), nil
+	}
+
+	totalHeadroom := resource.NewQuantity(0, resource.DecimalSI)
+	for _, r := range cra.regionMap {
+		headroom, err := r.GetHeadroom()
+		if err != nil {
+			return headroom, err
+		}
+		// FIXME: is it reasonable to simply add headroom together?
+		totalHeadroom.Add(headroom)
```

Comment:
Empty numa which not contain any dedicated_cores and share_cores is not considered here, I will fix it in the future PRs.